designs/NuGetExtensibility/NuGetExtensibilityServices.md

```diff
@@ -0,0 +1,161 @@
+If our services do not have any UI thread dependencies in any scenario, then they are [free threaded](https://github.com/microsoft/vs-threading/blob/b680de515d6210c4028598b988187c3aed33aabc/doc/cookbook_vs.md#how-do-i-effectively-verify-that-my-code-is-fully-free-threaded) and should be marked as such, both in the API documentation and in the [VS SDK analyzers](https://github.com/microsoft/VSSDK-Analyzers/blob/b0c5ddf01e137c6c938e9f37806210b9445f814b/src/Microsoft.VisualStudio.SDK.Analyzers.CodeFixes/build/AdditionalFiles/vs-threading.MembersRequiringMainThread.txt) to help customers use the correct calling context.
+
+Keep in mind that lots of our APIs won't need the UI thread unless we are initializing. See [VSPackageSourceProvider](https://github.com/NuGet/NuGet.Client/blob/f64621487c0b454eda4b98af853bf4a528bef72a/src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPackageSourceProvider.cs) for example. We will only initialize the VSSettings object when required. The VSSettings object might need to do a solution open check. It's super important to understand all scenarios before calling a service free threaded, because it will deadlock!
+
+In general refer to [VS Threading cookbook](https://github.com/microsoft/vs-threading/blob/b680de515d6210c4028598b988187c3aed33aabc/doc/cookbook_vs.md) over anything written in this document about Visual Studio threading.
+
+Refer to [IVsTestingExtension](https://github.com/NuGet/Entropy/tree/5215f3265663bd9c6f1c355b1d3190b803bd9600/IVsTestingExtension) for helpers for testing the threading patterns of your code. It also contains a ASYNC_FREETHREADED_CHECK option.
+
+### Exposing services for discoverability
+
+A boilerplate example for exposing services for discoverability by Visual Studio consumers.
+
+```cs
+namespace Microsoft.VisualStudio  // KEEP THIS NAMESPACE
+{
+    public static class NuGetServices  // CHANGE THIS CLASS NAME
+    {
+      private static ServiceRpcDescriptor Echo { get; } = new ServiceJsonRpcDescriptor(
+        new ServiceMoniker("Microsoft.VisualStudio.Echo", new Version("1.0")),
+        ServiceJsonRpcDescriptor.Formatters.UTF8,
+        ServiceJsonRpcDescriptor.MessageDelimiters.HttpLikeHeaders);
+
+      /// <summary>
+      /// Gets the <see cref="ServiceRpcDescriptor"/> for the echo service.
+      /// Use the <see cref="IEcho"/> interface for the client proxy for this service.
+      /// </summary>
+      public static ServiceRpcDescriptor Echo(this VisualStudioServices.VS2019_5Services svc) => Echo; // CHANGE SERVICE NAME
+    }
+}
+```
+
+### Deprecation of old services
+
+Most of the initial extensibility service work will involve migrating existing services.
+The old services should be marked as obsolete when the replacement ones are created.
+
+## Considerations
+
+* Should we use ValueTask over Task in our new extensibility interfaces?

```

Comment:
Keeping this as a question makes it sound like it's still open for discussion/decision. Even the text below makes it sound undecided. I suggest making the bullet point a statement that we are not using it, and the text below to say that although it was considered because it might have this advantage, it won't be used because of these problems.

Reply:
Makes sense, I'll fix it.

Reply:
fixed